server/src/main/java/org/opensearch/common/metadata/MetadataManager.java

```diff
@@ -0,0 +1,102 @@
+ * compatible open source license.
+ */
+
+package org.opensearch.common.metadata;
+
+import java.io.IOException;
+import java.util.Map;
+
+import org.apache.lucene.codecs.CodecUtil;
+import org.apache.lucene.store.BufferedChecksumIndexInput;
+import org.apache.lucene.store.ChecksumIndexInput;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+
+/**
+ * Manages Metadata by adding version, codec in header and checksum in footer.
+ * @param <T> The type of metadata to be read/written
+ */
+public class MetadataManager<T> {
+    // This can be updated to hold a parserFactory and get relevant parsers based on the metadata versions
+    private final MetadataParser<T> parser;
+    private final int currentVersion;
+    private final String codec;
+
+    /**
+     * @param parser parser to read/write metadata from T
+     * @param currentVersion latest supported version of the metadata
+     * @param codec: metadata codec
+     */
+    public MetadataManager(MetadataParser<T> parser, int currentVersion, String codec) {
+        this.parser = parser;
+        this.currentVersion = currentVersion;
+        this.codec = codec;
+    }
+
+    /**
+     * Reads metadata content from {@code indexInput} and parses the read content to {@link T}.
+     * Before reading actual content, verifies the header with relevant codec and version.
+     * After reading the actual content, verifies the checksum as well
+     * @param indexInput metadata file input stream
+     * @return metadata content parsed into {@link T}
+     */
+    public T readMetadata(IndexInput indexInput) throws IOException {
+        ChecksumIndexInput checksumIndexInput = new BufferedChecksumIndexInput(indexInput);
+        checkHeader(checksumIndexInput);
+        T metadata = this.parser.readContent(checksumIndexInput);
+        checkFooter(checksumIndexInput);
+        return metadata;
+    }
+
+    /**
+     * Writes metadata to file output stream {@code indexOutput}
+     * @param indexOutput file output stream which will store metadata content
+     * @param metadata metadata content.
+     *
+     * {@code metadata} argument type would be changed to {@link T} in future to reuse {@link MetadataManager} for multiple metadata types
+     */
+    public void writeMetadata(IndexOutput indexOutput, Map<String, String> metadata) throws IOException {
+        this.writeHeader(indexOutput);
+        this.parser.writeContent(indexOutput, metadata);
+        this.writeFooter(indexOutput);
+    }
+
+    /**
+     * Reads header from metadata file input stream containing {@code this.codec} and {@code this.currentVersion}.
+     * @param indexInput metadata file input stream
+     * @return header version found in the metadata file
+     */
+    private int checkHeader(IndexInput indexInput) throws IOException {
+        return CodecUtil.checkHeader(indexInput, this.codec, this.currentVersion, this.currentVersion);
```

Comment:
can we add a TODO here? The current version should probably not be dictating the min and max version for the already uploaded metadata. This again probably could come from a CurrentSupportedVersion input at the time of reading the metadata (`public T readMetadata(IndexInput indexInput)`).

Reply:
min and max both can be current if we only support a single version. But I agree we need to add support for multiple versions in MetadataManager. I was planning to do this as part of https://github.com/opensearch-project/OpenSearch/issues/6123 . currently i have refactored to provide a base to make versioning related changes in future

I'll add a TODO